--- ConicBundle-v0.3.11.src/Matrix/sparsmat.hxx	2017-05-04 17:02:59.000000000 -0400
+++ ConicBundle-v0.3.11.src/Matrix/sparsmat.hxx	2017-05-04 18:32:41.000000000 -0400
@@ -95,6 +95,14 @@ namespace CH_Matrix_Classes {
 
    */
 
+// The class below declares these as 'friend' functions,
+// but simultaneously attempts to provide default arguments for them.
+// That's not allowed, but we can declare them here first, with their default arguments.
+// See: http://stackoverflow.com/questions/23333949/the-correct-way-to-define-default-argument-for-a-friend-function-in-c
+class Sparsemat;
+Sparsemat& xbpeya(Sparsemat& x,const Sparsemat& y,Real alpha=1.,Real beta=0.,int ytrans=0);
+
+
 class Sparsemat: protected Memarrayuser
 {
     friend class Indexmatrix;
@@ -341,7 +349,7 @@ public:
   //@{
 
   ///returns x= alpha*y+beta*x, where y may be transposed (ytrans=1); if beta==0. then x is initialized to the correct size
-  friend Sparsemat& xbpeya(Sparsemat& x,const Sparsemat& y,Real alpha=1.,Real beta=0.,int ytrans=0);
+  friend Sparsemat& xbpeya(Sparsemat& x,const Sparsemat& y,Real alpha,Real beta,int ytrans);
   
   ///returns C=beta*C+alpha*A*B, where A and B may be transposed; C must not be equal to A and B; if beta==0. then C is initialized to the correct size
   friend Matrix& genmult(const Sparsemat& A,const Matrix& B,Matrix &C,
